_sources/configuration.txt
.. highlight:: none

Configuration
=============


Control File
------------

Configuration options and parameters in BAMM are specified in a *control file*,
a plain text file in which each line contains the name of the option or
parameter, an equal sign, and the value of the option or parameter::

    treefile = whaletree.tre
    runMCMC = 1

The path of the control file, relative to the directory in which ``bamm``
is called, is specified with the flag ``-c``::

    ./bamm -c divcontrol.txt


General Options and Parameters
------------------------------

The following describes the configuration options and parameters
that are required regardless of the specific model used.
For true or false values, 1 is used for true and 0 is used for false.
File paths are relative to the directory in which ``bamm`` is called.

General
.......

``modeltype``
    The type of model BAMM uses for analysis.
    If ``speciationextinction``, run with a speciation/extinction model.
    If ``trait,`` run with a phenotypic evolution model.

``treefile``
    The file path of the input tree (in Newick format).
    For speciation/extinction analyses, the tree must be ultrametric,
    fully bifurcating, and have unique tip names.

``runInfoFilename``
    The path of the file to output general information about the current run.

``sampleFromPriorOnly``
    If ``1``, run by sampling from the prior only
    (ignoring likelihood contribution to posterior).
    If ``0``, run the full analysis.

``autotune``
    Experimental option for tuning MCMC operators.

``runMCMC``
    If ``1``, run the MCMC sampler.
    If ``0``, just check to see if the data can be loaded correctly.

``simulatePriorShifts``
    If ``1``, simulate prior distribution of the number of shift events,
    given the hyperprior on the Poisson rate parameter.
    This is necessary to compute the Bayes factor.

``loadEventData``
    If ``1``, load a previous event state (event locations and parameter
    values) from a file. This allows a run to be continued even after a
    previous run has finished. The file format is identical to that of the
    default output file ``event_data.txt``. Only the data from the last
    generation will be processed if the file contains data for multiple
    generations. For example, to load the data from a previous run, copy the
    output event file (``event_data.txt``) to something like
    ``event_data_in.txt`` and use this file in the ``eventDataInfile`` option
    (below). Note that ``event_data.txt`` will be overwritten by the new run
    (unless you change its name in the configuration file), so if you'd like to
    keep those data, copy the file to a new one.

``eventDataInfile``
    The file path of the event data to be loaded (used only if ``loadEventData
    = 1``).

``initializeModel``
    If ``1``, initializes MCMC.
    If ``0``, just check parameter file and ensure that data can be read.

``seed``
    Seed for the random number generator.
    If ``-1``, the seed is obtained from the clock time.

``overwrite``
    If ``1``, overwrite output files if they already exist with identical
    filenames.
    If ``0``, do not run if output files already exist with identical filenames.

``validateEventConfiguration``
    If ``1``, rejects proposals that cause a branch and both of its direct
    descendants to have at least one event. Such event configuration may cause
    the parameters of the parent event to change to unrealistic values.
    If ``0``, no such proposals are immediately rejected.
    The default value is ``0``.


Priors
......

``poissonRatePrior``
    The rate parameter of the exponential prior on the rate parameter
    of the Poisson process. Smaller values favor greater numbers of distinct
    evolutionary regimes on the tree. Suggested values:
    ``poissonRatePrior = 1.0`` for smaller datasets (< 500 tips) or
    ``poissonRatePrior = 0.1`` or even ``0.2`` for large (5000+ tips).

MCMC Simulation
...............

``numberOfGenerations``
    Number of MCMC generations to run.

``mcmcWriteFreq``
    Frequency (in generations) at which to print MCMC details
    to ``mcmcOutfile``.

``eventDataWriteFreq``
    Frequency (in generations) at which to print event details
    to ``eventDataOutfile``.

``printFreq``
    Frequency (in generations) at which to print output to the screen.

``outName``
    If present (may be commented out), prefixes output files with the given
    text.

``mcmcOutfile``
    The path of the file to which to write the MCMC output.

``eventDataOutfile``
    The path of the file to which to write the raw event data.
    All of the results are contained in this file, and all branch-specific
    speciation rates, shift positions, marginal distributions, etc.,
    can be reconstructed from this output. See :ref:`bammtools`
    for more information on working with this output format.

``outputAcceptanceInfo``
    If ``1``, outputs whether each proposal was accepted.
    The number identifying the proposal matches the one in the code.
    The default value is ``0`` (i.e., do not output this information).

``acceptanceInfoFileName``
    The path of the file to which to write whether each proposal was accepted.
    ``outputAcceptedInfo`` must be set to ``1`` for this information to be
    written. The default value is ``acceptance_info.txt``.

``acceptanceResetFreq``
    Frequency in which to reset the acceptance information.
    The default value is ``1000``.

``updateEventLocationScale``
    Scale parameter for updating local moves of events on the tree.
    This defines the width of the sliding window proposal. This parameter
    is specified in units of "total tree depth" to minimize scale dependence. 
    Suppose you have a tree of age *T* (*T* is the time of the root node). Parameter 
    ``updateEventLocationScale`` is in units of T. A value of 0.05 means that the uniform
    distribution for event location changes has a width of 0.05T.

``updateEventRateScale``
    Scale parameter (proportional shrinking/expanding) for updating
    the rate parameter of the Poisson process.

``localGlobalMoveRatio``
    Ratio of local to global moves of events.

Metropolis Coupled MCMC
.......................

The Metropolis coupled MCMC implementation in BAMM is based on a
"global exchange scheme" for synchronization as described in
Altekar et al (2004) Parallel Metropolis coupled Markov chain
Monte Carlo for Bayesian phylogenetic inference.

``numberOfChains``
    Number of Markov chains to run. The default value is ``1``.

``deltaT``
    Temperature increment parameter. This value should be > 0.
    The temperature for the :math:`i`-th chain is calculated as
    :math:`1 / [1 + \Delta T \times (i - 1)]`.
    The default value is ``0.1``.

``swapPeriod``
    Number of generations in which to propose a chain swap.
    The default value is ``1000``.

``chainSwapFileName``
    Name of the file in which to output data about each chain swap proposal.
    The format of each line is
    ``[generation],[rank_1],[rank_2],[swap_accepted]``
    where ``[generation]`` is the generation when the swap proposal was made,
    ``[rank_1]`` and ``[rank_2]`` are the chains that were chosen
    (chain 1 is the coldest, chain 2 is the second coldest, and so on),
    and ``[swap_accepted]`` is whether the swap was made.
    The default value is ``chain_swap.txt``.


Parameter Update Rates
......................

``updateRateEventNumber``
    Relative frequency of MCMC moves that change the number of events,
    where the location of a new event is randomly picked across the whole tree.

``updateRateEventNumberForBranch``
    Relative frequency of MCMC moves that change the number of events,
    where the location of a new event is chosen by first
    picking a random branch, then a random location within that branch.
    This setting is hidden and set to 0 by default.
    If used (i.e., set to > 0), ``updateRateEventNumber`` should be set to 0.

``updateRateEventPosition``
    Relative frequency of MCMC moves that change the location of an event
    on the tree.

``updateRateEventRate``
    Relative frequency of MCMC moves that change the rate at which events occur.

``initialNumberEvents``
    Initial number of non-root processes.


Speciation/Extinction Model
---------------------------

The following describes the configuration options and parameters
that are specific to speciation/extinction analyses in BAMM.

General
.......

``useGlobalSamplingProbability``
    If ``1``, look for a global correction for incomplete sampling
    (globalSamplingProbability).
    If ``0``, look for a file that specifies clade-specific corrections
    for incomplete sampling (``sampleProbsFilename``).

``globalSamplingProbability``
    Percentage of total number of species sampled in the phylogeny
    (between 0 and 1).

``sampleProbsFilename``
    The path of a file containing clade-specific corrections for
    incomplete sampling.

Priors
......

``lambdaInitPrior``
    Prior on the inital lambda (rate parameter of the exponential distribution)
    for the speciation rate. Applies to all non-root events.
    
``lambdaShiftPrior``
    Prior on the the lambda shift parameter (standard deviation of the normal
    distribution) for the speciation rate. The mean of the distribution
    is fixed at zero, which is equal to a constant rate diversification process.
    Applies to non-root events.

``lambdaIsTimeVariablePrior``
    Prior on the time mode for the speciation rate.
    This prior is the probability that the speciation rate for a new event
    is time-variable (i.e., time-dependent) vs. constant through time.

``muInitPrior``
    Prior on the extinction rate (rate paramater of the exponential
    distribution). Applies to non-root events.
    
``segLength``
    The "grain" of the likelihood calculations. It approximates the
    continuous-time change in diversification rates by breaking each branch
    into a constant-rate diversification segments, with each segment equal to
    ``segLength``. The parameter is specified in units of total tree depth. If
    you have a tree of age T = 100, and set ``segLength = 0.05``, the segment
    size will be 5.  A branch of length 20 would thus have the exponential
    speciation-rate change approximated by 4 segments. If the value is greater
    than the branch length (e.g., ``segLength = 0.20`` in this case) BAMM will
    not break the branch into segments but use the mean rate across the entire
    branch.

MCMC Simulation
...............

``updateLambdaInitScale``
    Scale parameter for updating the initial speciation rate for each process.

``updateLambdaShiftScale``
    Scale parameter for the exponential change parameter for speciation.

``updateMuInitScale``
    Scale parameter for updating initial extinction rate for each process.

``minCladeSizeForShift``
    Allows you to constrain the location of possible rate-change events
    to occur only on branches with at least this many descendant tips.
    A value of ``1`` allows shifts to occur on all branches.

Starting Parameters
...................

``lambdaInit0``
    Initial speciation rate (at the root of the tree).

``lambdaShift0``
    Initial rate change parameter for speciation at the root.
    If ``0``, speciation rates will not change through time.
    A negative value implies decreasing rates through time.

``muInit0``
    Initial extinction rate at the root.

Parameter Update Rates
......................

``updateRateLambda0``
    Relative frequency of MCMC moves that change the initial speciation rate
    associated with an event.

``updateRateLambdaShift``
    Relative frequency of MCMC moves that change the exponential shift parameter
    of a speciation rate associated with an event.

``updateRateLambdaTimeMode``
    Relative frequency of MCMC moves that change whether the speciation rate
    for an event is time-variable (i.e., time-dependent)
    or constant through time.

``updateRateMu0``
    Relative frequency of MCMC moves that change the extinction rate for a given
    event.


Phenotypic Evolution Model
--------------------------

The following describes the configuration options and parameters
specific to the phenotypic evolution model in BAMM.
The parameter "beta" represents the rate of phenotypic evolution
at any point in time.

General
.......

``traitfile``
    The path to a file that contains the phenotypic trait data.
    Traits must be continuous characters.
    Each line must have a species name and the corresponding trait value,
    separated by a tab.
    A header row is **not** permitted.
    All species in the trait data file must be in the tree and vice versa.

MCMC Tuning
...........

``updateBetaInitScale``
    Scale operator for proportional shrinking/expanding move to update
    the initial phenotypic rate for rate regimes.

``updateNodeStateScale``
    Scale operator for sliding window move to update ancestral states
    at internal nodes.

``updateBetaShiftScale``
    Scale operator for sliding window move to update initial phenotypic rate.

Starting Parameters
...................

``betaInit``
    Initial value of the phenotypic evolutionary process at the root
    of the tree.

``betaShiftInit``
    Initial value of the exponential change parameter for the phenotypic
    evolutionary process (at the root of the tree).
    If ``0``, then the process has a constant rate.
    If negative, it implies decreasing rates through time.

Priors
......

``betaInitPrior``
    Parameter (rate) of the prior (exponential) on the inital phenotypic
    evolutionary rate associated with regimes, for non-root events.

``betaShiftPrior``
    Parameter (stdandard deviation) of the prior (normal) on the rate-change
    parameter for non-root events.

``betaIsTimeVariablePrior``
    Prior on the time mode for the phenotypic evolution rate.
    This prior is the probability that the phenotypic rate for a new event
    is time-variable (i.e., time-dependent) vs. constant through time.

``useObservedMinMaxAsTraitPriors``
    If ``1``, puts a uniform prior density on the distribution of ancestral
    character states, with upper and lower bonds determined by the min and max
    of the observed data.

``traitPriorMin``
    User-defined minimum value for the uniform density on the distribution
    of ancestral character states. Only used if
    ``useObservedMinMaxAsTraitPriors = 0``.

``traitPriorMax``
    User-defined maximum value for the uniform density on the distribution
    of ancestral character states. Only used if
    ``useObservedMinMaxAsTraitPriors = 0``.
    
Parameter Update Rates
......................

``updateRateBeta0``
    Relative frequency of moves that change the initial phenotypic rate
    associated with an event.

``updateRateBetaShift``
    Relative frequency of moves that change the exponential shift parameter
    of a phenotypic rate associated with an event.

``updateRateBetaTimeMode``
    Relative frequency of MCMC moves that change whether
    the phenotypic evolution rate for an event is time-variable
    (i.e., time-dependent) or constant through time.

``updateRateNodeState``
    Relative frequency of moves update the value of ancestral character stats.
    You have as many ancestral states as you have internal nodes in your tree,
    so there are a lot of parameters: this value should, in general,
    be substantially higher than the other parameter values

ChangeLog.txt
2014-06-09  Carlos Anderson  carlosjanderson@gmail.com

        Update: Example control files were updated for BAMM 2.0.0.

2014-06-05  Carlos Anderson  carlosjanderson@gmail.com

        Feature: Time-flip proposal was added for phenotypic evolution.

2014-06-02  Carlos Anderson  carlosjanderson@gmail.com

        Change: When the exponential rate parameter k > 0, use the symmetrical
        rate function of k < 0.

2014-04-25  Carlos Anderson  carlosjanderson@gmail.com

        Feature: Time-flip proposal was added for diversification model.

2014-04-16  Carlos Anderson  carlosjanderson@gmail.com

        Enhancement: Multiple Markov chains can run in parallel using OpenMP.

2014-04-09  Carlos Anderson  carlosjanderson@gmail.com

        Feature: More than one Markov chain can be created using Metropolis
        coupled MCMC.

2014-03-29  Carlos Anderson  carlosjanderson@gmail.com

        Enhancement: Internal node names are allowed in the tree file.

2014-03-12  Carlos Anderson  carlosjanderson@gmail.com

        Bug fix: Detection of negative branches was not done correctly.

2014-02-27  Dan Rabosky  dan.rabosky@gmail.com

        Change: Previously, numerical tolerance parameter MAX_E_PROB was
        defined as a macro in SpExModel.cpp with a default setting of 0.999.
        This is now an optional (usually not specified) parameter that can be
        changed in the control file or commmand line. This was necessary
        because a large dataset with highly incomplete sampling was getting
        non-finite initial likelihoods. The problem was solved after increasing
        the parameter to 0.9999. The new name is "ExtinctionProbMax".

2014-02-21  Carlos Anderson  carlosjanderson@gmail.com

        Change: If the simulate priors option is turned on, priors will be
        simulated after the MCMC has run, not before (as it did previously).

2014-02-14  Carlos Anderson  carlosjanderson@gmail.com

        Bug fix: In some systems, BAMM will not compile because the exit()
        method was not recognized. Fixed by including cstdlib in the relevant
        files.

2014-02-12  Carlos Anderson  carlosjanderson@gmail.com

        Bug fix: Prior values were not computed correctly for the root event.

        Feature: Any parameter in the control file can now be overriden by a
        command-line argument. The argument must start with "--", followed by
        the name of the parameter (as it is written in the control file),
        followed by a space, and finally followed by the new value.

2014-02-12  Dan Rabosky  dan.rabosky@gmail.com

        Feature: Separate priors may be specified for the root event.

        Bug fix: Log-likelihood was not updated after adding an event, causing
        the wrong log-likelihood to be reported.

2014-02-06  Carlos Anderson  carlosjanderson@gmail.com

        Bug fix: Error message "problem in BranchHistory::getLastEvent()"
        followed by a segmentation fault. Fixed by deleting the object pointed
        to by an iterator before erasing the iterator in the event collection.


src/Settings.cpp
#include <iostream>
#include <sstream>
#include <fstream>
#include <iomanip>
#include <vector>
#include <cctype>
#include <algorithm>
#include <cstdlib>

#include "Settings.h"
#include "Log.h"
#include "MatchPathSeparator.h"


Settings::Settings(const std::string& controlFilename,
    const std::vector<UserParameter>& commandLineParameters) :
    _commandLineParameters(commandLineParameters)
{
    readControlFile(controlFilename);

    // Get the model type
    std::string modelType;
    std::vector<UserParameter>::const_iterator it;
    for (it = _userParameters.begin(); it < _userParameters.end(); ++it) {
        if (it->first == "modeltype") {
            modelType = it->second;
            break;
        }
    }

    initializeGlobalSettings();

    // Initialize specific settings for model type
    if (modelType == "speciationextinction") {
        initializeSpeciationExtinctionSettings();
    } else if (modelType == "trait") {
        initializeTraitSettings();
    } else {
        exitWithErrorInvalidModelType();
    }

    // Re-assign parameters based on user values
    initializeSettingsWithUserValues();

    checkAllSettingsAreUserDefined();
    checkAllOutputFilesAreWriteable();
}


void Settings::readControlFile(const std::string& controlFilename)
{
    if (!fileExists(controlFilename)) {
        exitWithErrorNoControlFile();
    }

    std::ifstream controlStream(controlFilename.c_str());

    while (!controlStream.eof()) {
        std::string line;
        getline(controlStream, line, '\n');

        // Strip whitespace
        line.erase(std::remove_if(line.begin(), line.end(),
            (int(*)(int))isspace), line.end());

        // Strip comments
        std::istringstream lineStringStream(line);
        std::string lineWithoutComments;
        getline(lineStringStream, lineWithoutComments, '#');

        // Skip empty lines
        if (lineWithoutComments.size() == 0) {
            continue;
        }

        // Use second getline to split by '=' characters
        std::istringstream equalsStringStream(lineWithoutComments);
        std::vector<std::string> tokens;
        std::string token;
        while (getline(equalsStringStream, token, '=')) {
            tokens.push_back(token);
        }

        // Ensure input line is valid (two sides to an equal sign)
        if (tokens.size() != 2) {
            exitWithErrorInvalidLine(lineWithoutComments);
        }

        // Store parameter and its value
        _userParameters.push_back(UserParameter(tokens[0], tokens[1]));
    }
}


void Settings::initializeGlobalSettings()
{
    // General
    addParameter("modeltype", "speciationextinction");
    addParameter("treefile", "tree.txt");
    addParameter("sampleFromPriorOnly", "0", NotRequired);
    addParameter("runMCMC", "0");
    addParameter("loadEventData", "0", NotRequired);
    addParameter("eventDataInfile", "event_data_in.txt", NotRequired);
    addParameter("initializeModel", "0");
    addParameter("simulatePriorShifts", "1", NotRequired);
    addParameter("numberOfGenerations", "0");
    addParameter("seed", "-1", NotRequired);
    addParameter("validateEventConfiguration", "1", NotRequired);
    addParameter("checkUltrametric", "1", NotRequired);

    // MCMC tuning
    addParameter("updateEventLocationScale", "0.0");
    addParameter("updateEventRateScale", "0.0");
    addParameter("localGlobalMoveRatio", "0.0");

    // Metropolis-coupled MCMC
    addParameter("numberOfChains", "1", NotRequired);
    addParameter("deltaT", "0.1", NotRequired);
    addParameter("swapPeriod", "1000", NotRequired);
    addParameter("chainSwapFileName", "chain_swap.txt", NotRequired);

    // Priors
    addParameter("poissonRatePrior", "0.0");

    // Output
    addParameter("outName", "", NotRequired);
    addParameter("runInfoFilename", "run_info.txt", NotRequired);
    addParameter("mcmcOutfile", "mcmc_out.txt", NotRequired);
    addParameter("eventDataOutfile", "event_data.txt", NotRequired);
    addParameter("priorOutputFileName", "prior_probs.txt", NotRequired);

    addParameter("branchRatesWriteFreq", "0", NotRequired);
    addParameter("mcmcWriteFreq", "0");
    addParameter("eventDataWriteFreq", "0");

    addParameter("printFreq", "0");
    addParameter("overwrite", "0", NotRequired);
    addParameter("writeMeanBranchLengthTrees", "0", NotRequired);

    addParameter("acceptanceResetFreq", "1000", NotRequired);

    // Parameter update rates
    addParameter("updateRateEventNumber", "0.0");
    addParameter("updateRateEventNumberForBranch", "0.0", NotRequired);
    addParameter("updateRateEventPosition", "0.0");
    addParameter("updateRateEventRate", "0.0");
    addParameter("initialNumberEvents", "0");

    // Other (TODO: Need to add documentation for these)
    addParameter("autotune", "0", NotRequired);
    addParameter("outputAcceptanceInfo", "0", NotRequired);
    addParameter("acceptanceInfoFileName", "acceptance_info.txt", NotRequired);

    // TODO: New params May 30 2014, need documented
    addParameter("maxNumberEvents", "500", NotRequired);
    addParameter("priorSim_IntervalGenerations", "5000", NotRequired);
    addParameter("fastSimulatePrior_Generations", "5000000", NotRequired);
    addParameter("fastSimulatePrior_SampleFreq", "50", NotRequired);
    addParameter("fastSimulatePriorExperimental", "0", NotRequired);
    addParameter("fastSimulatePrior_BurnIn", "0.05", NotRequired);

    // maxNumberEvents = for fastSimulatePriorExperimental, maximum number of events...
    // priorSims_intervalGenerations = number of generations per model pair
    // fastSimulatePrior_Generations = sampling gens for simulation of prior
    // fastSimulatePrior_SampleFreq = sampling frequency for fastSimulation of prior
    // fastSimulatePrior_BurnIn = fraction of samples to be discarded as burnin.

}


void Settings::initializeSpeciationExtinctionSettings()
{
    // General
    addParameter("useGlobalSamplingProbability", "1");
    addParameter("globalSamplingFraction", "0.0");
    addParameter("sampleProbsFilename", "sample_probs.txt", NotRequired);

    // MCMC tuning
    addParameter("updateLambdaInitScale", "0.0");
    addParameter("updateMuInitScale", "0.0");
    addParameter("updateLambdaShiftScale", "0.0");
    addParameter("updateMuShiftScale", "0.0", NotRequired);
    addParameter("minCladeSizeForShift", "1", NotRequired);

    // Starting parameters
    addParameter("lambdaInit0", "0.0");
    addParameter("lambdaShift0", "0.0");
    addParameter("muInit0", "0.0");
    addParameter("muShift0", "0.0", NotRequired);

    // Priors
    addParameter("lambdaInitPrior", "0.0");
    addParameter("lambdaShiftPrior", "0.0");
    addParameter("muInitPrior", "0.0");
    addParameter("muShiftPrior", "1.0", NotRequired);
    addParameter("lambdaInitRootPrior", "-1.0", NotRequired);
    addParameter("lambdaShiftRootPrior", "-1.0", NotRequired);
    addParameter("muInitRootPrior", "-1.0", NotRequired);
    addParameter("muShiftRootPrior", "-1.0", NotRequired);
    addParameter("lambdaIsTimeVariablePrior", "0.5");
	addParameter("segLength", "0.0");

    // Output
    addParameter("lambdaOutfile", "lambda_rates.txt", NotRequired, Deprecated);
    addParameter("muOutfile", "mu_rates.txt", NotRequired, Deprecated);

    // Parameter update rates
    addParameter("updateRateLambda0", "0.0");
    addParameter("updateRateLambdaShift", "0.0");
    addParameter("updateRateMu0", "0.0");
    addParameter("updateRateMuShift", "0.0", NotRequired);
    addParameter("updateRateLambdaTimeMode", "0.0");

    // Maximum value of extinction probability on branch that will be tolerated:
    // to avoid numerical overflow issues (especially rounding to 1)
    addParameter("extinctionProbMax", "0.999", NotRequired);
}


void Settings::initializeTraitSettings()
{
    // General
    addParameter("traitfile", "traits.txt");

    // MCMC tuning
    addParameter("updateBetaInitScale", "0.0");
    addParameter("updateNodeStateScale", "0.0");
    addParameter("updateBetaShiftScale", "0.0");

    // Starting parameters
    addParameter("betaInit", "0.0");
    addParameter("betaShiftInit", "0.0");

    // Priors
    addParameter("betaInitPrior", "0.0");
    addParameter("betaShiftPrior", "0.0");
    addParameter("betaInitRootPrior", "-1.0", NotRequired);
    addParameter("betaShiftRootPrior", "-1.0", NotRequired);
    addParameter("betaIsTimeVariablePrior", "0.5");
    addParameter("useObservedMinMaxAsTraitPriors", "1");
    addParameter("traitPriorMin", "0.0", NotRequired);
    addParameter("traitPriorMax", "0.0", NotRequired);

    // Output
    addParameter("betaOutfile", "beta_rates.txt", NotRequired, Deprecated);
    addParameter("nodeStateOutfile", "node_state.txt", NotRequired);
    addParameter("nodeStateWriteFreq", "0", NotRequired);

    // Parameter update rates
    addParameter("updateRateBeta0", "0.0");
    addParameter("updateRateBetaShift", "0.0");
    addParameter("updateRateNodeState", "0.0");
    addParameter("updateRateBetaTimeMode", "0.0");
}


void Settings::addParameter(const std::string& name, const std::string& value,
    UserDefinedStatus userDefined, DeprecationStatus deprecated)
{
    _parameters.insert(Parameter(name,
        SettingsParameter(name, value, userDefined, deprecated)));
}


void Settings::initializeSettingsWithUserValues()
{
    std::vector<std::string> paramsNotFound;

    std::vector<UserParameter>::const_iterator userParamIt;
    for (userParamIt = _userParameters.begin();
        userParamIt != _userParameters.end(); ++userParamIt) {

        // Find the matching parameter to the user-specified parameter
        ParameterMap::iterator paramIt = _parameters.find((userParamIt->first));

        // If found, set the value of the parameter to the user's
        if (paramIt != _parameters.end()) {
            // Parameter should not be deprecated
            if ((paramIt->second).isDeprecated()) {
                exitWithErrorParameterIsDeprecated(paramIt->first);
            }

            // Parameter should not already be user-defined
            if ((paramIt->second).isUserDefined()) {
                exitWithErrorDuplicateParameter(paramIt->first);
            }

            (paramIt->second).setStringValue(userParamIt->second);
        } else {
            paramsNotFound.push_back(userParamIt->first);
        }
    }

    // Handle command-line arguments
    std::vector<UserParameter>::const_iterator cmdLineParamIt;
    for (cmdLineParamIt = _commandLineParameters.begin();
        cmdLineParamIt != _commandLineParameters.end(); ++cmdLineParamIt) {
        // Find the command-line parameter in known parameter list
        ParameterMap::iterator paramIt =
            _parameters.find((cmdLineParamIt->first));

        // If found, replace current parameter value with command-line value
        if (paramIt != _parameters.end()) {
            (paramIt->second).setStringValue(cmdLineParamIt->second);
        } else {
            log(Error) << "Command-line parameter " << cmdLineParamIt->first
                << " is not a known parameter.\n";
            std::exit(1);
        }
    }

    attachPrefixToOutputFiles();

    if (paramsNotFound.size() > 0) {
        exitWithErrorParametersNotFound(paramsNotFound);
    }
}


void Settings::attachPrefixToOutputFiles()
{
    // Get the prefix string value (in outName parameter)
    ParameterMap::const_iterator it = _parameters.find("outName");
    std::string prefix;
    if (it != _parameters.end()) {
        prefix = (it->second).value<std::string>();
    }

    // Create an array of the parameters that need to be prefixed
    std::string paramsToPrefix[NumberOfParamsToPrefix] =
        { "runInfoFilename",
          "mcmcOutfile",
          "eventDataOutfile",
          "nodeStateOutfile",
          "priorOutputFileName",
          "acceptanceInfoFileName",
          "chainSwapFileName",
          "lambdaOutfile",
          "muOutfile",
          "betaOutfile" };

    // Attach the prefix to each parameter
    ParameterMap::iterator paramIt;
    for (size_t i = 0; i < NumberOfParamsToPrefix; i++) {
        paramIt = _parameters.find(paramsToPrefix[i]);
        if (paramIt != _parameters.end()) {
            const std::string& param = (paramIt->second).value<std::string>();
            (paramIt->second).setStringValue(attachPrefix(prefix, param));
        }
    }
}


std::string Settings::attachPrefix
  (const std::string& prefix, const std::string& path) const
{
    if (prefix == "") {
        return path;
    }

    const std::string& dir = extractDir(path);
    const std::string& fileName = extractFileName(path);

    return dir + prefix + "_" + fileName;
}


std::string Settings::extractDir(const std::string& path) const
{
    // Starts from the beginning, stops when it finds the last path separator
    return std::string(path.begin(), std::find_if(path.rbegin(), path.rend(),
        MatchPathSeparator()).base());
}


std::string Settings::extractFileName(const std::string& path) const
{
    // Starts from the end and stops when it finds a path separator
    return std::string(std::find_if(path.rbegin(), path.rend(),
        MatchPathSeparator()).base(), path.end());
}


void Settings::checkAllSettingsAreUserDefined() const
{
    ParameterMap::const_iterator it;
    for (it = _parameters.begin(); it != _parameters.end(); ++it) {
        const SettingsParameter& parameter = it->second;
        if (!parameter.isDeprecated()) {
            if (!parameter.isUserDefined() && parameter.mustBeUserDefined()) {
                exitWithErrorUndefinedParameter(parameter.name());
            }
        }
    }
}


void Settings::checkAllOutputFilesAreWriteable() const
{
    if (!get<bool>("overwrite")) {
        if (anyOutputFileExists()) {
            exitWithErrorOutputFileExists();
        }
    }
}


bool Settings::anyOutputFileExists() const
{
    // Global output files
    if (fileExists(get("runInfoFilename")) ||
        fileExists(get("mcmcOutfile"))     ||
        fileExists(get("eventDataOutfile"))) {
        return true;
    }

    if (get<bool>("writeMeanBranchLengthTrees")) {
        // Speciation/extinction output files
        if (get("modeltype") == "speciationextinction") {
            if (fileExists(get("lambdaOutfile")) ||
                fileExists(get("muOutfile"))) {
                return true;
            }

        // Trait output files
        } else if (get("modeltype") == "trait") {
            if (fileExists(get("betaOutfile"))) {
                return true;
            }
        }
    }

    return false;
}


bool Settings::fileExists(const std::string& filename) const
{
    std::ifstream inFile(filename.c_str());
    return inFile.good();
}


void Settings::printCurrentSettings(std::ostream& out) const
{
    int ppw = 29;

    log(Message, out) << "Current parameter settings:\n";

    ParameterMap::const_iterator it;
    for (it = _parameters.begin(); it != _parameters.end(); ++it) {
        log(Message, out) << std::right << std::setw(ppw) <<
            it->first << "    " << (it->second).value<std::string>() << "\n";
    }

    log(Message, out) << std::flush;
}


void Settings::exitWithErrorNoControlFile() const
{
    log(Error) << "Specified control file does not exist.\n"
               << "Check that the file is in the specified location.\n";
    std::exit(1);
}


void Settings::exitWithErrorInvalidLine(const std::string& line) const
{
    log(Error) << "Invalid input line in control file.\n"
               << "Problematic line includes <<" << line << ">>\n";
    std::exit(1);
}


void Settings::exitWithErrorUndefinedParameter(const std::string& name) const
{
    log(Error) << "Parameter " << name << " is undefined.\n"
               << "Fix by giving the parameter a value in the control file.\n";
    std::exit(1);
}


void Settings::exitWithErrorInvalidModelType() const
{
    log(Error) << "Invalid type of analysis.\n"
               << "Fix by setting modeltype as speciationextinction or trait\n";
    std::exit(1);
}


void Settings::exitWithErrorParametersNotFound
    (const std::vector<std::string>& paramsNotFound) const
{
    log(Error) << "One or more parameters from the control file does not\n"
        << "correspond to valid model parameters. Make sure that you are\n"
        << "running the correct version of BAMM or check that the following\n"
        << "parameters are spelled correctly:\n\n";

    std::vector<std::string>::const_iterator it;
    for (it = paramsNotFound.begin(); it != paramsNotFound.end(); ++it) {
        log() << std::setw(30) << *it << std::endl;
    }

    std::exit(1);
}

void Settings::exitWithErrorParameterIsDeprecated
    (const std::string& param) const
{
    log(Error) << "Parameter " << param << " has been deprecated. Fix by\n"
        << "removing this parameter or use the appropriate version of BAMM.\n";
    std::exit(1);
}


void Settings::exitWithErrorDuplicateParameter(const std::string& param) const
{
    log(Error) << "Duplicate parameter " << param << ".\n"
               << "Fix by removing duplicate parameter in control file.\n";
    std::exit(1);
}


void Settings::exitWithErrorOutputFileExists() const
{
    log(Error) << "Analysis is set to not overwrite files.\n"
               << "Fix by removing or renaming output file(s),\n"
               << "or set \"overwrite = 1\" in the control file.\n";
    std::exit(1);
}
